### PR TITLE
dialects: (arith) add addi constant propagation

### PR DESCRIPTION
**Pull request by alexarice**
Moves constants to the right for an addi operation, similar to what #3557 did for muli. Also replaces `AddImmediateZero` by `AddiIdentityRight`, which removes constant zeros on the right of an addition instead of the left of an addition like the previous pattern did.

Updates interactive tests (and driveby fixes the constant zero being given the SSA value name %two in this test).

### CODE REVIEW

**Review by n-io — APPROVED** [review]

File: xdsl/transforms/canonicalization_patterns/arith.py [hunk 1]
```diff
@@ -9,15 +9,30 @@
...
+        if (lhs := const_evaluate_operand(op.lhs)) is None:
+            return
+        if (rhs := const_evaluate_operand(op.rhs)) is None:
+            # Swap inputs if lhs is constant and rhs is not
+            rewriter.replace_matched_op(arith.AddiOp(op.rhs, op.lhs))
+            return
+
+        assert isinstance(op.result.type, IntegerType | IndexType)
+        rewriter.replace_matched_op(
+            arith.ConstantOp.from_int_and_width(lhs + rhs, op.result.type)
```
Comment (superlopuh) [inline xdsl/transforms/canonicalization_patterns/arith.py:34]: overflow?

Comment (superlopuh) [reply xdsl/transforms/canonicalization_patterns/arith.py:34]: I guess we'll crash in the short term and will be able to deal with it then. I just thought I'd raise it now

Comment (alexarice) [reply xdsl/transforms/canonicalization_patterns/arith.py:34]: This probably effects the previous PR too

Comment (alexarice) [reply xdsl/transforms/canonicalization_patterns/arith.py:34]: I think this is the equivalent part of mlir https://github.com/llvm/llvm-project/blob/9ad09b2930ef2e95bf8772c91f623881d1c14733/mlir/lib/Dialect/Arith/IR/ArithOps.cpp#L325

I can't see any mention of overflow flags or wrapping.

File: tests/filecheck/dialects/arith/canonicalize.mlir [hunk 2]
```diff
@@ -108,3 +108,19 @@ func.func @test_const_var_const() {
...
+
+%zero_plus = arith.addi %c0, %a : i32
+%plus_zero = arith.addi %a, %c0 : i32
+
+// CHECK: "test.op"(%a, %a) {"identity addition check"} : (i32, i32) -> ()
+"test.op"(%one_times, %times_one) {"identity addition check"} : (i32, i32) -> ()
+
+// CHECK: %plus_const = arith.addi %a, %c2 : i32
+%plus_const = arith.addi %c2, %a : i32
+"test.op"(%plus_const) : (i32) -> ()
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/filecheck/dialects/arith/canonicalize.mlir:122]: Unless I'm missing something, it looks like this is a copy/paste from above, but is not necessary

**Review by superlopuh — APPROVED** [review]

File: tests/filecheck/dialects/arith/canonicalize.mlir (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexarice) [reply tests/filecheck/dialects/arith/canonicalize.mlir:122]: It was also a copy paste

**Comment by alexarice** [comment]
Have fixed the test and resolved the conflict

**Review by compor — APPROVED** [review]

**Pull request merged**